freqtrade/optimize/backtesting.py

```diff
@@ -350,8 +351,70 @@ def _get_close_rate(self, sell_row: Tuple, trade: LocalTrade, sell: SellCheckTup
         else:
             return sell_row[OPEN_IDX]
 
+    def _get_adjust_trade_entry_for_candle(self, trade: LocalTrade, row: Tuple
+                                           ) -> LocalTrade:
+
+        current_profit = trade.calc_profit_ratio(row[OPEN_IDX])
+
+        stake_amount = strategy_safe_wrapper(self.strategy.adjust_trade_position,
+                                             default_retval=None)(
+            pair=trade.pair, trade=trade, current_time=row[DATE_IDX].to_pydatetime(),
+            current_rate=row[OPEN_IDX], current_profit=current_profit)
+
+        # Check if we should increase our position
+        if stake_amount is not None and stake_amount > 0.0:
+            return self._execute_trade_position_change(trade, row, stake_amount)
```

Comment:
You'll need to be careful here, as this is not properly emulating live, potentially causing quite different results.

In live, `execute_entry()` will be called to increase a position.
That method calls several strategy callbacks itself, namely
* `custom_entry_price`
* `custom_stake_amount`
* `confirm_trade_entry`

for both new trades, as well as for DCA "increases".
To be clear - i think that's correct behavior (although the `custom_stake_amount` might be unnecessary) - as it allows the user to determine a buy price - as well as "back out" last second (once you see the price you'd get, for example).

In backtesting, we're calling the 3 above mentioned callbacks too - but ONLY for the first trade entry - not for any subsequent DCA orders. Depending on the logic in these methods (maybe there is none) - this can become a problem as you'll no longer be able to properly backtest the proper workings of these methods in DCA situations.

Reply:
I've reworked that into _enter_trade to reduce code duplication and make sure we are calling the callbacks correctly.
custom_entry_price is called, custom_stake_amount and confirm_trade_entry are skipped for both bot and backtest.
custom_stake_amount because adjust_trade_position already returns stakes and confirm_trade_entry because it already send the stake only if it wants the trade, so essentially confirming it at the same time. Also adjust_trade_position has the actually pre-existing trade / previous orders available so it has more data to make a qualified decision.